### ISSUE TITLE
Add length normalization option to DPO Trainer

### ISSUE THREAD

**Issue by ggbetz** [huggingface/trl#2964]
### Feature request

DPO Trainer, [unlike CPO Trainer](https://github.com/huggingface/trl/blob/69ad852e5654a77f1695eb4c608906fe0c7e8624/trl/trainer/cpo_trainer.py#L781) with ipo or simpo loss, does not seem to support length normalization.

Could you consider to add a length normalization option to the dpo trainer?

### Motivation

DPO training with sigmoid loss and length normalization is part of the Tülu 3 recipe ([report](https://arxiv.org/pdf/2411.15124v4), Sec 5.4.1).

So we currently cannot use trl to follow Tülu cookbook.

❤️ Thanks so much for building trl!

### Your contribution

TBD

**Comment by 1485840691-eng** [comment]
There are several works that address length normalization in DPO. I would propose https://arxiv.org/pdf/[2409.06411](https://arxiv.org/pdf/2409.06411) that address length normalization from re-weighting log probabilities and gets better metrics than DPO and SimPO

<img width="2112" height="1801" alt="Image" src="https://github.com/user-attachments/assets/d94d973c-ca30-4353-bb17-bc9859c9b6ca" />

If possible, I would propose to work out a PR for it. Any early feedback, please let me know. Thanks
@qgallouedec @LeonEricsson @kashif

**Comment by qgallouedec** [comment]
Thanks we would definitely welcome a PR

**Comment by 1485840691** [comment]
I think LD-DPO is already supported

        ld_alpha (`float` or `None`, *optional*, defaults to `None`):
            α parameter from the [LD-DPO paper](https://huggingface.co/papers/2409.06411), which controls the weighting
            of the verbose token log-probabilities in responses. If `None`, no weighting is applied to the verbose
            part, and the loss is equivalent to the standard DPO loss. The paper recommends setting `ld_alpha` between
            `0.0` and `1.0`.

**Comment by BrownianNotion** [comment]
hey @qgallouedec, just wondering how come this issue was closed? I'm also interested in length-normalised DPO given it's still being used in the OLMo series today. was it because @1485840691 was primarily interested in LD-DPO? Thank you!

**Comment by BrownianNotion** [comment]
The implementation should hopefully be straightforward if we create a separate loss type for this (eg. `sigmoid_length_norm`) and instead of trying to support mixing this with other loss types (which may normalise in other ways anyway, so this wouldn't make sense). We should also assert/check that divergence type is standard DPO/reverse KL.

The code should be the same as the `ipo` [loss type](https://github.com/huggingface/trl/blob/main/trl/trainer/dpo_trainer.py/#L1258-L1269), except the final line where `per_sequence_loss` is calculated would be replaced by `per_sequence_loss = -F.logsigmoid(self.beta * delta_score)`, where `delta_score` is calculated in the same way as `ipo_delta`.

Happy to try this and open a PR if that sounds good to you!
